File: javascript/cobra_core.js
/**
 * COBRA: COines BRidge Access — Core Protocol Layer (JavaScript/WebSerial)
 *
 * Implements the COINES V3 binary protocol over WebSerial API.
 * See ../core/PROTOCOL.md for the language-agnostic specification.
 *
 * Usage:
 *   const bridge = new CobraBridge();
 *   await bridge.connect();
 *   const chipId = await bridge.i2cRead(0x14, 0x00, 1);
 *   console.log(`Chip ID: 0x${chipId[0].toString(16).padStart(2, '0')}`);
 *   bridge.disconnect();
 */

import {
    HEADER, TYPE_GET, TYPE_SET,
    CMD_I2C_READ, CMD_I2C_WRITE,
    CMD_SPI_READ, CMD_SPI_WRITE,
    CMD_GET_BOARD_INFO, CMD_SET_VDD, CMD_SET_VDDIO,
    STATUS_OK,
    I2C_SPEED_400K, I2C_SPEED_1M,
    SPI_SPEED_5MHZ, SPI_SPEED_10MHZ,
    SPI_MODE_0, SPI_MODE_3,
} from './cobra_constants.js';


export class CobraBridge {
    constructor() {
        this.port = null;
        this.reader = null;
        this.writer = null;
        this.readableStream = null;
        this.writableStream = null;
    }

    // ── Connection ──────────────────────────────────────────────────────

    async connect(baudRate = 115200) {
        this.port = await navigator.serial.requestPort();
        await this.port.open({ baudRate });
        this.readableStream = this.port.readable.getReader();
        this.writableStream = this.port.writable.getWriter();
    }

    async disconnect() {
        if (this.readableStream) { await this.readableStream.cancel(); this.readableStream = null; }
        if (this.writableStream) { await this.writableStream.close(); this.writableStream = null; }
        if (this.port) { await this.port.close(); this.port = null; }
    }

    get connected() { return this.port !== null; }

    // ── Low-Level Protocol (core/PROTOCOL.md §1) ─────────────────────

    static checksum(data) {
        let xor = 0;
        for (const b of data) xor ^= b;
        return xor;
    }

    buildPacket(type, command, payload = new Uint8Array(0)) {
        const length = payload.length;
        const frame = new Uint8Array(5 + length + 1);
        frame[0] = HEADER;
        frame[1] = type;
        frame[2] = command;
        frame[3] = length & 0xFF;
        frame[4] = (length >> 8) & 0xFF;
        frame.set(payload, 5);
        let xor = 0;
        for (let i = 0; i < 5 + length; i++) xor ^= frame[i];
        frame[5 + length] = xor;
        return frame;
    }

    async sendPacket(type, command, payload = new Uint8Array(0)) {
        const pkt = this.buildPacket(type, command, payload);
        await this.writableStream.write(pkt);
    }

    async receivePacket(timeout = 2000) {
        const deadline = Date.now() + timeout;
        const buf = [];

        // 1. Wait for 0xAA
        while (Date.now() < deadline) {
            const { value } = await this.readableStream.read();
            if (!value) continue;
            for (const b of value) {
                buf.push(b);
                if (b === HEADER && buf.length === 1) break;
            }
            if (buf.length === 1 && buf[0] === HEADER) break;
        }
        if (buf.length === 0 || buf[0] !== HEADER) throw new Error('No header');

        // Helper: read N bytes
        const readN = async (n) => {
            while (buf.length < n + 1) { // +1 for header already read
                const { value } = await this.readableStream.read();
                if (value) for (const b of value) buf.push(b);
            }
            return buf.slice(1, 1 + n); // skip header
        };

        // 2. Read header rest
        const headerBytes = await readN(4);
        const ptype = headerBytes[0];
        const command = headerBytes[1];
        const length = headerBytes[2] | (headerBytes[3] << 8);

        // 3. Read payload + checksum
        while (buf.length < 5 + length + 2) { // header + 4 + payload + checksum
            const { value } = await this.readableStream.read();
            if (value) for (const b of value) buf.push(b);
        }

        const payload = new Uint8Array(buf.slice(5, 5 + length));
        const receivedXor = buf[5 + length];

        // 5. Verify checksum
        const frame = new Uint8Array(buf.slice(0, 5 + length));
        const expectedXor = CobraBridge.checksum(frame);
        if (expectedXor !== receivedXor) {
            throw new Error(`Checksum mismatch: 0x${expectedXor.toString(16)} vs 0x${receivedXor.toString(16)}`);
        }

        // 6. Extract status and data
        const status = payload.length > 0 ? payload[0] : STATUS_OK;
        const data = payload.length > 1 ? payload.slice(1) : new Uint8Array(0);

        // Clear buffer
        buf.length = 0;

        return { type: ptype, command, status, data };
    }

    async transact(type, command, payload = new Uint8Array(0), timeout = 2000) {
        await this.sendPacket(type, command, payload);
        const resp = await this.receivePacket(timeout);
        return { status: resp.status, data: resp.data };
    }

    // ── I2C Operations (core/PROTOCOL.md §3) ──────────────────────────

    async i2cWrite(devAddr, regAddr, data, speed = I2C_SPEED_400K) {
        const payload = new Uint8Array([devAddr, speed, regAddr, data.length, ...data]);
        const { status } = await this.transact(TYPE_SET, CMD_I2C_WRITE, payload);
        return status;
    }

    async i2cRead(devAddr, regAddr, length, speed = I2C_SPEED_400K) {
        const payload = new Uint8Array([devAddr, speed, regAddr, length]);
        const { status, data } = await this.transact(TYPE_GET, CMD_I2C_READ, payload);
        if (status !== STATUS_OK) throw new Error(`I2C read failed: 0x${status.toString(16)}`);
        return data;
    }

    // ── SPI Operations (core/PROTOCOL.md §4) ──────────────────────────

    async spiWrite(csLine, regAddr, data, speed = SPI_SPEED_5MHZ, mode = SPI_MODE_0) {
        const payload = new Uint8Array([csLine, speed, mode, regAddr, data.length, ...data]);
        const { status } = await this.transact(TYPE_SET, CMD_SPI_WRITE, payload);
        return status;
    }

    async spiRead(csLine, regAddr, length, speed = SPI_SPEED_5MHZ, mode = SPI_MODE_0) {
        const payload = new Uint8Array([csLine, speed, mode, regAddr, length]);
        const { status, data } = await this.transact(TYPE_GET, CMD_SPI_READ, payload);
        if (status !== STATUS_OK) throw new Error(`SPI read failed: 0x${status.toString(16)}`);
        return data;
    }

    // ── Board Control (core/PROTOCOL.md §5) ──────────────────────────

    async getBoardInfo() {
        const { status, data } = await this.transact(TYPE_GET, CMD_GET_BOARD_INFO);
        if (status !== STATUS_OK) throw new Error('Board info failed');
        return { raw: data, boardId: data[0] | (data[1] << 8) };
    }

    async setVdd(voltageMv) {
        const lo = voltageMv & 0xFF, hi = (voltageMv >> 8) & 0xFF;
        const { status } = await this.transact(TYPE_SET, CMD_SET_VDD, new Uint8Array([lo, hi]));
        return status;
    }

    async setVddio(voltageMv) {
        const lo = voltageMv & 0xFF, hi = (voltageMv >> 8) & 0xFF;
        const { status } = await this.transact(TYPE_SET, CMD_SET_VDDIO, new Uint8Array([lo, hi]));
        return status;
    }
}

File: python/cobra_core.py
"""
COBRA: COines BRidge Access — Core Protocol Layer (Python)

Implements the COINES V3 binary protocol over USB-Serial.
See ../core/PROTOCOL.md for the language-agnostic specification.

Usage:
    from cobra_core import CobraBridge

    bridge = CobraBridge(port='/dev/ttyUSB0')
    bridge.connect()
    chip_id = bridge.i2c_read(dev_addr=0x14, reg_addr=0x00, length=1)
    print(f"Chip ID: 0x{chip_id[0]:02X}")
    bridge.disconnect()
"""

import struct
import time
from typing import Optional

from cobra_constants import (
    HEADER, TYPE_GET, TYPE_SET,
    CMD_I2C_READ, CMD_I2C_WRITE,
    CMD_SPI_READ, CMD_SPI_WRITE,
    CMD_GET_BOARD_INFO, CMD_SET_VDD, CMD_SET_VDDIO,
    STATUS_OK,
    I2C_SPEED_400K, I2C_SPEED_1M,
    SPI_SPEED_5MHZ, SPI_SPEED_10MHZ,
    SPI_MODE_0, SPI_MODE_3,
)


class CobraBridge:
    """
    Synchronous COINES V3 bridge over USB-Serial.

    Provides send_packet / receive_packet primitives plus
    convenience methods for I2C and SPI register access.
    """

    def __init__(self, port: str = '/dev/ttyUSB0', baudrate: int = 115200, timeout: float = 2.0):
        self.port = port
        self.baudrate = baudrate
        self.timeout = timeout
        self._ser = None

    # ── Connection ────────────────────────────────────────────────────────

    def connect(self):
        """Open the serial port."""
        import serial
        self._ser = serial.Serial(
            port=self.port,
            baudrate=self.baudrate,
            timeout=self.timeout,
            bytesize=8, parity='N', stopbits=1,
        )
        time.sleep(0.1)
        self._ser.reset_input_buffer()
        self._ser.reset_output_buffer()

    def disconnect(self):
        """Close the serial port."""
        if self._ser and self._ser.is_open:
            self._ser.close()
        self._ser = None

    @property
    def connected(self) -> bool:
        return self._ser is not None and self._ser.is_open

    # ── Low-Level Protocol (see core/PROTOCOL.md §1) ─────────────────────

    @staticmethod
    def _checksum(data: bytes) -> int:
        """XOR checksum of all bytes in data."""
        xor = 0
        for b in data:
            xor ^= b
        return xor

    def build_packet(self, ptype: int, command: int, payload: bytes = b'') -> bytes:
        """
        Build a COINES V3 packet.

        Implements the packet building algorithm from core/PROTOCOL.md §1:
            frame = [0xAA, type, command, length_lo, length_hi, payload...]
            checksum = XOR_reduce(frame)
            return frame + [checksum]
        """
        length = len(payload)
        frame = struct.pack('<BBBB', HEADER, ptype, command, length & 0xFF)
        frame += struct.pack('B', (length >> 8) & 0xFF)
        frame += payload
        xor = self._checksum(frame)
        return frame + struct.pack('B', xor)

    def send_packet(self, ptype: int, command: int, payload: bytes = b'') -> None:
        """Build and send a COINES V3 packet."""
        pkt = self.build_packet(ptype, command, payload)
        if not self.connected:
            raise ConnectionError("Serial port not connected")
        self._ser.write(pkt)
        self._ser.flush()

    def receive_packet(self, timeout: Optional[float] = None) -> tuple:
        """
        Read and parse a COINES V3 response packet.

        Implements the packet parsing algorithm from core/PROTOCOL.md §1.
        Returns (ptype, command, status, data) tuple.
        """
        if not self.connected:
            raise ConnectionError("Serial port not connected")

        t_out = timeout or self.timeout
        deadline = time.time() + t_out

        # 1. Wait for header 0xAA
        while time.time() < deadline:
            b = self._ser.read(1)
            if b == b'\xAA':
                break
        else:
            raise TimeoutError("No response header received")

        # 2. Read type, command, length_lo, length_hi
        header_rest = self._ser.read(4)
        if len(header_rest) < 4:
            raise TimeoutError("Incomplete response header")
        ptype, command, length_lo, length_hi = struct.unpack('<BBBB', header_rest)
        length = length_lo | (length_hi << 8)

        # 3. Read payload
        payload = b''
        if length > 0:
            payload = self._ser.read(length)
            if len(payload) < length:
                raise TimeoutError(f"Incomplete payload: expected {length}, got {len(payload)}")

        # 4. Read checksum
        xor_byte = self._ser.read(1)
        if len(xor_byte) < 1:
            raise TimeoutError("Missing checksum byte")

        # 5. Verify checksum
        frame = struct.pack('<BBBB', HEADER, ptype, command, length_lo)
        frame += struct.pack('B', length_hi) + payload
        expected_xor = self._checksum(frame)
        if expected_xor != xor_byte[0]:
            raise ValueError(
                f"Checksum mismatch: expected 0x{expected_xor:02X}, got 0x{xor_byte[0]:02X}"
            )

        # 6. Extract status (first byte) and data
        status = payload[0] if len(payload) > 0 else STATUS_OK
        resp_data = payload[1:] if len(payload) > 1 else b''

        return (ptype, command, status, resp_data)

    def transact(self, ptype: int, command: int, payload: bytes = b'',
                 timeout: Optional[float] = None) -> tuple:
        """Send a packet and wait for the response. Returns (status, data)."""
        self.send_packet(ptype, command, payload)
        _, _, status, resp_data = self.receive_packet(timeout)
        return status, resp_data

    # ── I2C Operations (see core/PROTOCOL.md §3) ─────────────────────────

    def i2c_write(self, dev_addr: int, reg_addr: int, data: bytes,
                  speed: int = I2C_SPEED_400K) -> int:
        """Write data to an I2C register. Returns status byte."""
        payload = struct.pack('<BBBB', dev_addr, speed, reg_addr, len(data)) + data
        status, _ = self.transact(TYPE_SET, CMD_I2C_WRITE, payload)
        return status

    def i2c_read(self, dev_addr: int, reg_addr: int, length: int,
                 speed: int = I2C_SPEED_400K) -> bytes:
        """Read data from an I2C register. Returns bytes read."""
        payload = struct.pack('<BBBB', dev_addr, speed, reg_addr, length)
        status, resp_data = self.transact(TYPE_GET, CMD_I2C_READ, payload)
        if status != STATUS_OK:
            raise IOError(f"I2C read failed with status 0x{status:02X}")
        return resp_data

    # ── SPI Operations (see core/PROTOCOL.md §4) ─────────────────────────

    def spi_write(self, cs_line: int, reg_addr: int, data: bytes,
                  speed: int = SPI_SPEED_5MHZ, mode: int = SPI_MODE_0) -> int:
        """Write data to an SPI register. Returns status byte."""
        payload = struct.pack('<BBBBB', cs_line, speed, mode, reg_addr, len(data)) + data
        status, _ = self.transact(TYPE_SET, CMD_SPI_WRITE, payload)
        return status

    def spi_read(self, cs_line: int, reg_addr: int, length: int,
                 speed: int = SPI_SPEED_5MHZ, mode: int = SPI_MODE_0) -> bytes:
        """Read data from an SPI register. Returns bytes read."""
        payload = struct.pack('<BBBBB', cs_line, speed, mode, reg_addr, length)
        status, resp_data = self.transact(TYPE_GET, CMD_SPI_READ, payload)
        if status != STATUS_OK:
            raise IOError(f"SPI read failed with status 0x{status:02X}")
        return resp_data

    # ── Board Control (see core/PROTOCOL.md §5) ──────────────────────────

    def get_board_info(self) -> dict:
        """Get board identification information."""
        status, data = self.transact(TYPE_GET, CMD_GET_BOARD_INFO)
        if status != STATUS_OK:
            raise IOError(f"Get board info failed with status 0x{status:02X}")
        info = {'raw': data}
        if len(data) >= 2:
            info['board_id'] = struct.unpack('<H', data[0:2])[0]
        if len(data) >= 6:
            info['software_ver'] = f"{data[2]}.{data[3]}"
            info['hardware_ver'] = f"{data[4]}.{data[5]}"
        return info

    def set_vdd(self, voltage_mv: int) -> int:
        """Set VDD voltage in millivolts (0 = off)."""
        status, _ = self.transact(TYPE_SET, CMD_SET_VDD, struct.pack('<H', voltage_mv))
        return status

    def set_vddio(self, voltage_mv: int) -> int:
        """Set VDDIO voltage in millivolts (0 = off)."""
        status, _ = self.transact(TYPE_SET, CMD_SET_VDDIO, struct.pack('<H', voltage_mv))
        return status